File: spherical_harmonic_two_grain.m
function f=spherical_harmonic_two_grain(B0, susc, a, sep, alpha, L, debug_mag)

mu0 = 4*pi*1e-07;
mu = (1+susc)*mu0;
H0mag = B0/(mu0);  % Applied magnetic field, A/m
sep=sep*a;

R1=[0 0 sep]';
R2=[0 0 0]';

alpha = deg2rad(alpha); % change angle into radians
H0 =[H0mag*sin(alpha) 0 H0mag*cos(alpha)]'; % A/m

% parallel and perpendicular components of the applied magnetic field
H_perp=H0(1);
H_prll=H0(3);

%% Code to solve for coefficients (using paper)
for m=0:1
    % Creating the L X L matrices 
    X=zeros(L,L);
    Delta_m=zeros(L,L);
    Gamma_m=zeros(L,L);
    for i=1:L
        for j=1:L
            % X matrix
            if i==j
                X(i,j)= i*(mu/mu0) + i +1;
            end
            % Delta and Gamma matrix (paper: nchoosek(i+j,j+m))
            Delta_m(i,j)= ((-1)^(i+m))*(i*(mu/mu0)-i)*nchoosek(i+j,j-m)*...
                (a^(2*i+1))/(sep^(i+j+1));
            Gamma_m(i,j)= ((-1)^(i+j))*Delta_m(i,j);
        end
    end
    % 2LX2L matrix
    Am=zeros(2*L);
    Am(1:L,1:L)=X;
    Am(L+1:end,1:L)=Gamma_m;
    Am(1:L,L+1:end)=Delta_m;
    Am(L+1:end,L+1:end)=X;
    
    % q vector
    qm=zeros(L,1);
    if m==0
        qm(1)=-H_prll*(a^3)*(1-mu/mu0);
    elseif m==1
        qm(1)=H_perp*(a^3)*(1-mu/mu0);
    end
    % 2L Q vector
    Qm=[qm;qm];
    %Solve linear equations
    Beta_m=Am\Qm;
    Beta1_m=Beta_m(1:L);
    Beta2_m=Beta_m(L+1:2*L);
    if m==0
        Beta1_0=Beta1_m;
        Beta2_0=Beta2_m;
    elseif m==1
        Beta1_1=Beta1_m;
        Beta2_1=Beta2_m;
    end
end

%% Computing the Magnetic Field
% Create 3D spherical mesh
dang = pi/180;
inc = dang/2:dang:pi+dang/2;
az = dang/2:dang:2*pi+dang/2;
dr=a/100;
r1=a-dr:dr:a+dr;
[theta,phi,R] = meshgrid(inc,az,r1);


Hr=0;
Hth=0;
Hphi=0;

for l=1:L
    for m=0:1
        Hrs=0;
        Hths=0;
        
        for s=m:L
            Psm=legendre(s,cos(theta));
            Ps1m=legendre(s+1,cos(theta));
            if s~=0
                Psm=reshape(Psm(m+1,:,:,:),size(R));
            end
            Ps1m=reshape(Ps1m(m+1,:,:,:),size(R));
            %compute derivative of the associated legendre function
            dPsm=((m-s-1).*Ps1m + (s+1).*cos(theta).*Psm)./(-sin(theta));
            % R component
            Hrs=Hrs+ (-1)^(s+m) * nchoosek(l+s,s+m)*s.*(R.^(s-1)).*Psm/...
                (sep^(l+s+1));
            % Theta component
            Hths=Hths+ (-1)^(s+m) * nchoosek(l+s,s+m).*(R.^(s-1)).*dPsm./...
                ((sep^(l+s+1)));
        end
        Plm=legendre(l,cos(theta));
            Pl1m=legendre(l+1,cos(theta));
            Plm=reshape(Plm(m+1,:,:,:),size(R));
            Pl1m=reshape(Pl1m(m+1,:,:,:),size(R));
            
            %compute derivative of the associated legendre function
            dPlm=((m-l-1).*Pl1m + (l+1).*cos(theta).*Plm)./(-sin(theta));
            
        if m==0
            % Theta Component
            Hth=Hth+...
                (Beta1_0(l).*dPlm./((R.^(l+2))) + Beta2_0(l)*Hths).*cos(m*phi);
            % R Component
            Hr=Hr+...
                ((l+1)*Beta1_0(l).*Plm./(R.^(l+2)) -...
                Beta2_0(l)*Hrs).*cos(m*phi);
        elseif m==1
            % Theta Component
            Hth=Hth+...
                (Beta1_1(l).*dPlm./((R.^(l+2))) + Beta2_1(l)*Hths).*cos(m*phi);
            % R Component
            Hr=Hr+...
                ((l+1)*Beta1_1(l).*Plm./(R.^(l+2)) -...
                Beta2_1(l)*Hrs).*cos(m*phi);
        end
    end
end
%Phi component
for l=1:L
    Hphis=0;
    for s=1:L
        Ps1=legendre(s,cos(theta));
        Ps1=reshape(Ps1(2,:,:,:),size(R));
        Hphis=Hphis+ (-1)^(s+1) * nchoosek(l+s,s+1).*(R.^(s-1)).*Ps1...
            ./(sin(theta).*(sep^(l+s+1)));
    end
    Pl1=legendre(l,cos(theta));
    Pl1=reshape(Pl1(2,:,:,:),size(R));
    Hphi=Hphi+...
                (Beta1_1(l).*Pl1./(sin(theta).*(R.^(l+2))) +...
                Beta2_1(l)*Hphis).*sin(phi);
end
Hth=-Hth;
%% plot Magnetic Field
%Hmat = sqrt(Hr.^2+Hth.^2+Hphi.^2);
size_R=size(R);

if debug_mag==1
    x=R.*cos(phi).*sin(theta);
    y=R.*sin(phi).*sin(theta);
    z=R.*cos(theta);

    H_tot_mag=zeros(size_R);
for i=1:size_R(1)
    for j=1:size_R(2)
        for k=1:size_R(3)
            ph=az(i);
            th=inc(j);
            %H_part_sph=[Hr(i,j,k);Hphi(i,j,k);Hth(i,j,k)];
            %transformation matrix
                pre=[sin(th)*cos(ph), cos(th)*cos(ph), -sin(ph);...
                    sin(th)*sin(ph), cos(th)*sin(ph),  cos(ph);...
                    cos(th),         -sin(th),         0];
                post=pre';
            H0_sph=post*H0;
            H_sph=[Hr(i,j,k);Hth(i,j,k);Hphi(i,j,k)] + H0_sph;
            H_tot_mag(i,j,k)=norm(H_sph);
        end
    end
end

figure;
Ang=1;
pc=pcolor(squeeze(z(Ang,:,:))./a,squeeze(x(Ang,:,:))./a,squeeze(H_tot_mag(Ang,:,:))); set(pc, 'EdgeColor', 'none');
colormap('hot');
xlim([-2 2]); ylim([-1.5 1.5]);
title('|H|');
xlabel('x');
ylabel('y');
colorbar;
axis equal;
grid on;
end
%% Formulating the Maxwell Stress Tensor in Spherical Coordinates
f=0;
for i=1:size_R(1)
    if i==1 || i==size_R(1)
        p=1;
    elseif mod(i,2)~=0
        p=2;
    else
        p=4;
    end
    for j=1:size_R(2)
        if j==1 || j==size_R(2)
            q=1;
        elseif mod(j,2)~=0
            q=2;
        else
            q=4;
        end
            ph=az(i);
            th=inc(j);
            %transformation matrix
                pre=[sin(th)*cos(ph), cos(th)*cos(ph), -sin(ph);...
                    sin(th)*sin(ph), cos(th)*sin(ph),  cos(ph);...
                    cos(th),         -sin(th),         0];
                post=pre';
            H0_sph=post*H0;
            H_sph=[Hr(i,j,2);Hth(i,j,2);Hphi(i,j,2)] + H0_sph;
            H_cart=pre*H_sph;
            h=norm(H_cart);
            T_cart=mu0*(H_cart*H_cart' - 0.5*h^2*eye(3,3));
            rn_hat=[sin(th)*cos(ph),sin(th)*sin(ph),cos(th)]';
            f=f+a*a*T_cart*rn_hat*sin(th)*p*q;                  
    end
end
f=f*dang*dang/9;
end

File: spherical_harmonic_two_grain.py
import numpy as np
from scipy.special import comb
from scipy.special import lpmn
# import matplotlib.pyplot as plt

@np.vectorize
def lpmn_arr(m, n, x):
    return lpmn(m, n, x)[0][-1, -1]
@np.vectorize
def d_lpmn_arr(m, n, x):
    return lpmn(m, n, x)[1][-1, -1]

def spherical_harmonic_two_grain(B0, susc, a, sep, alpha, L):
    mu0= 4*np.pi*1e-07
    mu = (a+susc)*mu0
    H0mag = B0/mu0 #Applied magnetic field, A/m
    sep = sep*a

    R1=np.array([[0], [0], [sep]])
    R2=np.array([[0], [0], [0]])

    alpha= np.deg2rad(alpha) #change angle into radians
    H0= np.array([[H0mag*np.sin(alpha)], [0], [H0mag*np.cos(alpha)]]) # A/m

    # parallel and perpendicular components of the applied magnetic field
    H_perp=H0[0]
    H_prll=H0[2]

    # Code to solve for coefficents (using paper)
    for m in range(2):
        # Creating the L X L matrices
        X= np.zeros((L,L))
        Delta_m= np.zeros((L,L))
        Gamma_m= np.zeros((L,L))
        for i in range(L):
            for j in range(L):
                # X Matrix
                if i==j:
                    X[i,j]=(i+1)*(mu/mu0) + (i+1) + 1
                # Delta and Gamma matrix
                Delta_m[i,j]= ((-1)**((i+1)+m))*((i+1)*(mu/mu0)-(i+1))*comb(i+1+j+1, j+1-m, exact=True)*(a**(2*(i+1)+1))/(sep**(i+1+j+1+1))
                Gamma_m[i,j]= ((-1)**(i+1+j+1))*Delta_m[i,j]
                # 2L X 2L matrix
        Am=np.zeros((2*L,2*L))
        Am[0:L,0:L]=X
        Am[L:,0:L]=Gamma_m
        Am[0:L,L:]=Delta_m
        Am[L:,L:]=X

        # q vector
        qm= np.zeros((L,1))
        if m==0:
            qm[0,0]=-H_prll*(a**3)*(1-mu/mu0)
        elif m==1:
            qm[0,0]=H_perp*(a**3)*(1-mu/mu0)
        
        # 2L Q vector
        Qm=np.append(qm, qm)

        #solve linear equations
        Beta_m=np.linalg.solve(Am,Qm)
        Beta1_m=Beta_m[0:L]
        Beta2_m=Beta_m[L:]
        if m==0:
            Beta1_0=Beta1_m
            Beta2_0=Beta2_m
        elif m==1:
            Beta1_1=Beta1_m
            Beta2_1=Beta2_m

    ## Computing Magnetic Field

    #Create a 3D spherical mesh
    dang= np.pi/180.0
    inc= np.arange(dang/2, np.pi + dang, dang)
    az= np.arange(dang/2, 2*np.pi + dang, dang)
    dr= a/100.0
    r1= np.arange(a-dr,a+dr,dr)
    theta, phi, R=np.meshgrid(inc,az,r1)

    #Convert the spherical mesh to cartesian
    x= np.multiply(R,np.multiply(np.cos(phi),np.sin(theta)))
    y= np.multiply(R,np.multiply(np.sin(phi),np.sin(theta)))
    z= np.multiply(R,np.cos(theta))

    #define size parameters
    size_R=np.shape(R) #size of all elements in 3D space
    size_H=np.append(size_R,L) #size for 4D mag arrays

    Hr_L=np.zeros(size_H)
    Hth_L=np.zeros(size_H)
    Hphi_L=np.zeros(size_H)

    Hr=0
    Hth=0
    Hphi=0

    for l in np.arange(1,L+1):
        for m in range(2):
            Hrs=0
            Hths=0
            for s in np.arange(m,L+1):
                Psm=lpmn_arr(m,s,np.cos(theta))
                dPsm=d_lpmn_arr(m,s,np.cos(theta))*(-np.sin(theta))
                # R component
                additional=((-1)**(s+m))*s*np.multiply(np.power(R,s-1),Psm)/ (sep**(l+s+1))
                Hrs=Hrs + additional*comb(l+s,s+m, exact=True)
                # Theta component
                Hths=Hths + ((-1)**(s+m)) * comb(l+s,s+m, exact=True)*np.multiply(np.power(R,(s-1)),dPsm)/((sep**(l+s+1)))

            Plm=lpmn_arr(m,l,np.cos(theta))
            dPlm=d_lpmn_arr(m,l,np.cos(theta))*(-np.sin(theta))
            if m==0:
                # R Component
                Hr=Hr + np.multiply(((l+1)*Beta1_0[l-1]*np.divide(Plm,np.power(R,(l+2))) -  Beta2_0[l-1]*Hrs),np.cos(m*phi))
                # Theta component
                Hth=Hth + np.multiply((Beta1_0[l-1]*np.divide(dPlm,np.power(R,(l+2))) + Beta2_0[l-1]*Hths),np.cos(m*phi))
            elif m==1:
                # R Component
                Hr=Hr + np.multiply(((l+1)*Beta1_1[l-1]*np.divide(Plm,np.power(R,(l+2))) -  Beta2_1[l-1]*Hrs),np.cos(m*phi))
                # Theta component
                Hth=Hth + np.multiply((Beta1_1[l-1]*np.divide(dPlm,np.power(R,(l+2))) + Beta2_1[l-1]*Hths),np.cos(m*phi))
        Hr_L[:,:,:,l-1]=Hr
        Hth_L[:,:,:,l-1]=-Hth

    # Phi compenent
    for l in np.arange(1,L+1):
        Hphis=0
        for s in np.arange(1,L+1):
            Ps1=lpmn_arr(1,s,np.cos(theta))
            Hphis=Hphis + (-1)**(s+1) *comb(l+s,s+1, exact=True)*np.divide(np.multiply(np.power(R,s-1),Ps1),np.sin(theta))/(sep**(l+s+1))
        Pl1=lpmn_arr(1,l,np.cos(theta))
        # Hphi=Hphi + np.multiply((Beta1_1[l-1]*np.divide(Pl1,np.multiply(np.sin(theta),np.power(R,l+2))) + Beta2_1[l-1]*Hphis),np.sin(phi))
        Hphi_L[:,:,:,l-1]=Hphi

    Hth=-Hth

    # Formulating the Maxwell Stress Tensor in Spherical Coordinates
    f_L=np.zeros((3,1,L))
    for l in np.arange(1,L+1):
        f=0
        for i in range(size_R[0]):
            if i==0 or i==(size_R[0]-1):
                p=1
            elif np.mod(i,2)==0:
                p=2
            else:
                p=4
            for j in range(size_R[1]):
                if j==0 or j==size_R[1]-1:
                    q=1
                elif np.mod(j,2)==0:
                    q=2
                else:
                    q=4
                ph=az[i]
                th=inc[j]
                # Transformation matrix
                pre= np.array([[np.sin(th)*np.cos(ph), np.cos(th)*np.cos(ph), -np.sin(ph)],
                                [np.sin(th)*np.sin(ph), np.cos(th)*np.sin(ph),  np.cos(ph)],
                                [np.cos(th), -np.sin(th),  0]])
                post=np.transpose(pre)
                H0_sph=np.matmul(post,H0)
                H_sph=np.array([[Hr_L[i,j,1,l-1]],[Hth_L[i,j,1,l-1]],[Hphi_L[i,j,1,l-1]]]) + H0_sph
                H_cart=np.matmul(pre,H_sph)
                h=np.linalg.norm(H_cart)
                T_cart=mu0*(np.matmul(H_cart,np.transpose(H_cart)) - 0.5*(h**2)*np.eye(3))
                rn_hat=np.array([[np.sin(th)*np.cos(ph)],[np.sin(th)*np.sin(ph)],[np.cos(th)]])
                f=f+ a*a*np.sin(th)*p*q*np.matmul(T_cart,rn_hat)
        f=f*dang*dang/9.0
        f_L[:,:,l-1]=f  
    return(f)
# #plot z component of force vs L
# if debug_f_L==1:
#     plt.figure()
#     plt.plot(np.arange(1,L+1),f_L[2,0,:])
#     plt.title("Z component of force vs L (multipoles)")
#     plt.xlabel('L (multipoles)')
#     plt.ylabel('f_z (N)')
#     plt.grid(True)
#     plt.show()